### DISCUSSION TITLE
Encrypted Git Hosting

### DISCUSSION THREAD

**Discussion by vHanda** [GitJournal/GitJournal#851] [Feature Requests]
There are currently no open encrypted Git Hosting providers. Keybase was the only option, but they aren't open source. Perhaps GitJournal could start providing encrypted hosting, it would also make the project more sustainable.

The current plan is to create a new git remote handler, which operates over https. Each git object would be encrypted with a common passphrase before sending it to the server. This way the server would be storing opaque blobs. The files would not be encrypted at rest, and this way all the standard git commands will continue to work. This approach might work for individual repos, but sharing a passphrase doesn't scale with multiple users.

If anyone has ideas on how to implement encrypted git hosting, which does not encrypt the files at rest and would work with multiple users and keys, please let me know.

**Comment by chmac** [comment]
There are a few existing approaches, each with some drawbacks

* https://github.com/spwhitton/git-remote-gcrypt
  * - Every push is effectively a "force push", which is dangerous
  * + Works with any standard git remote
* https://github.com/AGWA/git-crypt
  * - Focused on encrypting only some files
  * + Uses smudge / clean filters, so supports any remote
* https://git-annex.branchable.com/encryption/
  * Would be interesting to see if GitJournal could interoperate with git-annex
* https://github.com/shadowhand/git-encrypt/tree/legacy
  * Deprecated, but could be interesting for inspiration

**Comment by clach04** [comment]
Alternative idea add an option for encrypted files. I.e. support both plain text and encrypted, kind of like the https://github.com/AGWA/git-crypt suggestion above.

This would still mean plain text files are open to being read on a remote server if the user did not choose to encrypt those specific files. Unclear if a history of encrypted files would leave then vulnerable to key analysis attacks (I'm not a crypto expert :-().

**Comment by cornelius** [comment]
What about https://foks.pub/? Would that be an option?

**Reply by vHanda** [reply]
I didn't know about folks. I just took a look and the fact that the client is written in go does make it easier to integrate as GitJournal currently uses go-git.

This definitely sounds interesting and more doable!
